Use secure temp directory for remote sync (#10911)

t-h-e opened the pull request:
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Explain large or complex code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" )
Fixes https://github.com/keepassxreboot/keepassxc/issues/10911

## Screenshots
[TIP]:  # ( Do not include screenshots of your actual database! )


## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )


## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ Bug fix (non-breaking change that fixes an issue)

t-h-e committed:
Fix use secure temp directory for remote sync (#10911)

t-h-e commented:
@hifi can you have a look at this?

hifi commented on src/gui/remote/RemoteHandler.cpp:
```diff
@@ -143,3 +134,27 @@ RemoteHandler::RemoteResult RemoteHandler::upload(const QString& filePath, const
     });
 }
+
+QString RemoteHandler::getTempFileLocation()
```
There's zero error checking done in this function. Both `QDir::mkdir` and `QFile::setPermissions` can return false and the existence of the directory isn't tested prior to creating it regardless how unlikely a collision is.

It's fine for the whole function to just fail if something fails and return an error to the user because it's very unlikely to fail so no retries are necessary.

t-h-e replied:
I have added checks to fail in case `QDir::mkdir` and `QFile::setPermissions` return false, although it is unlikely that either of them will return false. `QDir::mkdir` will also return false in case the directory already exists, so no additional existence check is needed.

t-h-e committed:
Add error handling

in case directory could not be created or permissions could not be set

t-h-e committed:
Update translations

droidmonkey commented on src/gui/remote/DatabaseSettingsWidgetRemote.cpp:
```diff
@@ -177,24 +177,27 @@ void DatabaseSettingsWidgetRemote::testDownload()
     if (!result.success) {
         m_ui->messageWidget->showMessage(tr("Download failed with error: %1").arg(result.errorMessage),
                                          MessageWidget::Error);
+        remoteHandler->cleanup(result.filePath);
```
Just call cleanup (or do the cleanup) in the destructor of the RemoteHandler class.

t-h-e replied:
Done. See https://github.com/keepassxreboot/keepassxc/pull/10914/commits/2c037385800b04104e25357f655ce2297f1f54be
This required a little refactor, so that only a single instance of `RemoteHandler` is used per sync.

t-h-e committed:
Use single instance of remote handler per sync

Pro:
- allows use of destructor to clean up
- remote handler can store file location and reuse it

Con:
- upload can not be called without previously calling download
  - currently not needed

droidmonkey commented:
Regarding the inability to call upload as a standalone thing, RemoteParams could include the "target" file path for using upload or download functions, if set the m_temporaryFileLocation variable is ignored. Another option is to just reintroduce the filepath variable to upload, but as an optional second parameter.

t-h-e commented:
> Regarding the inability to call upload as a standalone thing, RemoteParams could include the "target" file path for using upload or download functions, if set the m_temporaryFileLocation variable is ignored. Another option is to just reintroduce the filepath variable to upload, but as an optional second parameter.

I think it is fine as is as we currently have no standalone upload. In case, we want to add an upload only functionality, this is definitely the way to go.

droidmonkey commented on src/gui/DatabaseWidget.cpp:
```diff
@@ -1134,17 +1135,17 @@ void DatabaseWidget::syncDatabaseWithLockedDatabase(const QString& filePath, con
     if (result.success && !params->uploadCommand.isEmpty()) {
         emit updateSyncProgress(75, tr("Uploading..."));
-        result = remoteHandler->upload(result.filePath, params);
+        result = m_remoteHandler->upload(params);
```
Need to check for nullptr

t-h-e replied:
Done

t-h-e committed:
Handle possible nullptr exception
